File: scripts/listing_vetoes.py
"""Listing vetoes (Hide / Park) — Cockroach-backed map + pure desk apply helpers."""
from __future__ import annotations

import os
import sys
from datetime import datetime, timezone
from typing import Protocol

STATUS_HIDDEN = "hidden"
STATUS_PARKED = "parked"
VALID_STATUSES = frozenset({STATUS_HIDDEN, STATUS_PARKED})

DDL = """
CREATE TABLE IF NOT EXISTS listing_vetoes (
  item_id BIGINT NOT NULL PRIMARY KEY,
  status TEXT NOT NULL,
  updated_at TIMESTAMPTZ NOT NULL DEFAULT now()
);
"""

UPSERT_SQL = """
INSERT INTO listing_vetoes (item_id, status, updated_at)
VALUES (%(item_id)s, %(status)s, %(updated_at)s)
ON CONFLICT (item_id) DO UPDATE SET
  status = EXCLUDED.status,
  updated_at = EXCLUDED.updated_at
"""

DELETE_SQL = "DELETE FROM listing_vetoes WHERE item_id = %s"
LOAD_SQL = "SELECT item_id, status FROM listing_vetoes"
LOAD_HIDDEN_SQL = "SELECT item_id FROM listing_vetoes WHERE status = 'hidden'"


def _item_id(row_or_id) -> int | None:
    if isinstance(row_or_id, dict):
        raw = row_or_id.get("id", row_or_id.get("item_id"))
    else:
        raw = row_or_id
    if raw is None:
        return None
    try:
        return int(raw)
    except (TypeError, ValueError):
        return None


def is_hidden(vetoes: dict, item_id) -> bool:
    iid = _item_id(item_id)
    if iid is None:
        return False
    return vetoes.get(iid) == STATUS_HIDDEN or vetoes.get(str(iid)) == STATUS_HIDDEN


def is_parked(vetoes: dict, item_id) -> bool:
    iid = _item_id(item_id)
    if iid is None:
        return False
    return vetoes.get(iid) == STATUS_PARKED or vetoes.get(str(iid)) == STATUS_PARKED


def _status_for(vetoes: dict, item_id) -> str | None:
    iid = _item_id(item_id)
    if iid is None:
        return None
    return vetoes.get(iid) or vetoes.get(str(iid))


def apply_to_finds(rows: list, vetoes: dict, *, mode: str = "active") -> list:
    """Filter/tag/sort finds by veto map.

    mode:
      active  — omit hidden; tag parked; parked after active (default)
      parked  — only parked rows
      hidden  — only hidden rows
      all     — all rows with veto_status when set; parked after non-parked
    """
    out = []
    for row in rows:
        st = _status_for(vetoes, row)
        if mode == "active":
            if st == STATUS_HIDDEN:
                continue
        elif mode == "parked":
            if st != STATUS_PARKED:
                continue
        elif mode == "hidden":
            if st != STATUS_HIDDEN:
                continue
        elif mode == "all":
            pass
        else:
            raise ValueError(f"unknown veto mode: {mode}")
        tagged = dict(row)
        if st:
            tagged["veto_status"] = st
        else:
            tagged.pop("veto_status", None)
        out.append(tagged)

    def sort_key(r):
        st = r.get("veto_status")
        if st == STATUS_PARKED:
            return 1
        if st == STATUS_HIDDEN:
            return 2
        return 0

    # Stable: active, then parked, then hidden
    return sorted(out, key=sort_key)


def apply_to_bundles(rows: list, vetoes: dict, *, mode: str = "active") -> list:
    """Strip hidden members; drop bundle if <2 items remain; tag/sort parked."""
    if mode not in ("active", "parked", "hidden", "all"):
        raise ValueError(f"unknown veto mode: {mode}")

    out = []
    for bundle in rows:
        items = list(bundle.get("items") or [])
        kept_items = []
        for it in items:
            st = _status_for(vetoes, it)
            if mode == "active" and st == STATUS_HIDDEN:
                continue
            if mode == "parked" and st == STATUS_HIDDEN:
                continue
            if mode == "hidden" and st != STATUS_HIDDEN:
                continue
            tagged = dict(it)
            if st:
                tagged["veto_status"] = st
            kept_items.append(tagged)

        if mode == "hidden":
            if not kept_items:
                continue
        elif len(kept_items) < 2:
            continue
        elif mode == "parked" and not any(
            it.get("veto_status") == STATUS_PARKED for it in kept_items
        ):
            continue

        row = dict(bundle)
        row["items"] = kept_items
        listing_sum = 0.0
        for it in kept_items:
            try:
                listing_sum += float(it.get("price") or 0)
            except (TypeError, ValueError):
                pass
        row["listing_sum"] = listing_sum
        extra = row.get("checkout_extra_ron")
        if extra is not None:
            try:
                row["checkout_total"] = listing_sum + float(extra)
            except (TypeError, ValueError):
                pass

        any_parked = any(it.get("veto_status") == STATUS_PARKED for it in kept_items)
        any_hidden = any(it.get("veto_status") == STATUS_HIDDEN for it in kept_items)
        if mode == "hidden" or (mode == "all" and any_hidden and not any_parked):
            row["veto_status"] = STATUS_HIDDEN
        elif any_parked:
            row["veto_status"] = STATUS_PARKED
        else:
            row.pop("veto_status", None)
        out.append(row)

    def sort_key(r):
        st = r.get("veto_status")
        if st == STATUS_PARKED:
            return 1
        if st == STATUS_HIDDEN:
            return 2
        return 0

    return sorted(out, key=sort_key)


def item_is_hidden(item_id, hidden_ids: set) -> bool:
    """True when item_id is in the bot suppress set (hidden vetoes only)."""
    if item_id is None:
        return False
    try:
        return int(item_id) in hidden_ids
    except (TypeError, ValueError):
        return False


def filter_scored_rows(rows: list, hidden_ids: set) -> list:
    """Drop scored rows whose listing id is hidden."""
    if not hidden_ids:
        return list(rows)
    out = []
    for row in rows:
        item = row.get("item") if isinstance(row, dict) else None
        iid = (item or {}).get("id") if item else row.get("id")
        if item_is_hidden(iid, hidden_ids):
            continue
        out.append(row)
    return out


def filter_items(items: list, hidden_ids: set) -> list:
    """Drop raw Vinted item dicts whose id is hidden."""
    if not hidden_ids:
        return list(items)
    return [it for it in items if not item_is_hidden(it.get("id"), hidden_ids)]


class VetoStore(Protocol):
    def set_status(self, item_id: int, status: str) -> None: ...
    def clear(self, item_id: int) -> None: ...
    def load_map(self) -> dict[int, str]: ...
    def load_hidden_ids(self) -> set[int]: ...
    def close(self) -> None: ...


class MemoryVetoStore:
    def __init__(self) -> None:
        self._map: dict[int, str] = {}

    def set_status(self, item_id: int, status: str) -> None:
        status = str(status)
        if status not in VALID_STATUSES:
            raise ValueError(f"invalid veto status: {status}")
        self._map[int(item_id)] = status

    def clear(self, item_id: int) -> None:
        self._map.pop(int(item_id), None)

    def load_map(self) -> dict[int, str]:
        return dict(self._map)

    def load_hidden_ids(self) -> set[int]:
        return {iid for iid, st in self._map.items() if st == STATUS_HIDDEN}

    def close(self) -> None:
        return None


class NullVetoStore:
    def set_status(self, item_id: int, status: str) -> None:
        return None

    def clear(self, item_id: int) -> None:
        return None

    def load_map(self) -> dict[int, str]:
        return {}

    def load_hidden_ids(self) -> set[int]:
        return set()

    def close(self) -> None:
        return None


class PsycopgVetoStore:
    def __init__(self, conn) -> None:
        self._conn = conn

    def set_status(self, item_id: int, status: str) -> None:
        status = str(status)
        if status not in VALID_STATUSES:
            raise ValueError(f"invalid veto status: {status}")
        now = datetime.now(timezone.utc)
        with self._conn.cursor() as cur:
            cur.execute(
                UPSERT_SQL,
                {"item_id": int(item_id), "status": status, "updated_at": now},
            )
        self._conn.commit()

    def clear(self, item_id: int) -> None:
        with self._conn.cursor() as cur:
            cur.execute(DELETE_SQL, (int(item_id),))
        self._conn.commit()

    def load_map(self) -> dict[int, str]:
        with self._conn.cursor() as cur:
            cur.execute(LOAD_SQL)
            return {int(r[0]): str(r[1]) for r in cur.fetchall()}

    def load_hidden_ids(self) -> set[int]:
        with self._conn.cursor() as cur:
            cur.execute(LOAD_HIDDEN_SQL)
            return {int(r[0]) for r in cur.fetchall()}

    def close(self) -> None:
        try:
            self._conn.close()
        except Exception:
            pass


def ensure_schema(conn) -> None:
    with conn.cursor() as cur:
        cur.execute(DDL)
    conn.commit()


def database_url() -> str | None:
    return (
        os.environ.get("DATABASE_URL")
        or os.environ.get("COCKROACH_DATABASE_URL")
        or ""
    ).strip() or None


def open_store() -> VetoStore:
    url = database_url()
    if not url:
        return NullVetoStore()
    try:
        import psycopg

        conn = psycopg.connect(url, connect_timeout=10)
        ensure_schema(conn)
        return PsycopgVetoStore(conn)
    except Exception as e:
        print(f"listing_vetoes: DB unavailable, using null store: {e}", file=sys.stderr)
        return NullVetoStore()

File: src/server/listingVetoes.ts
// @ts-nocheck
import fs from "node:fs"
import path from "node:path"
import os from "node:os"
import pg from "pg"
import { resolveFamily } from "#/server/tasteLearning"

/**
 * Listing vetoes (Remove / Park / Bought) — Cockroach map + pure desk apply helpers.
 * Mirrors python/listing_vetoes.py.
 */

const STATUS_REMOVED = "removed"
const STATUS_PARKED = "parked"
const STATUS_BOUGHT = "bought"
const STATUS_HIDDEN_LEGACY = "hidden"
const VALID = new Set([STATUS_REMOVED, STATUS_PARKED, STATUS_BOUGHT])
const VALID_MODES = new Set(["active", "parked", "bought", "all"])

const ENRICHMENT_FIELDS = [
  "hunt_name",
  "hunt_family",
  "brand",
  "size",
  "price_ron",
  "value_band",
  "deal_score",
  "title",
]

const DDL = `
CREATE TABLE IF NOT EXISTS listing_vetoes (
  item_id BIGINT NOT NULL PRIMARY KEY,
  status TEXT NOT NULL,
  updated_at TIMESTAMPTZ NOT NULL DEFAULT now(),
  hunt_name TEXT NULL,
  hunt_family TEXT NULL,
  brand TEXT NULL,
  size TEXT NULL,
  price_ron DOUBLE PRECISION NULL,
  value_band TEXT NULL,
  deal_score INT NULL,
  title TEXT NULL
);
`

const MIGRATE_HIDDEN_SQL =
  "UPDATE listing_vetoes SET status = 'removed' WHERE status = 'hidden'"

const ALTER_COLUMNS_SQL = [
  "ALTER TABLE listing_vetoes ADD COLUMN IF NOT EXISTS hunt_name TEXT NULL",
  "ALTER TABLE listing_vetoes ADD COLUMN IF NOT EXISTS hunt_family TEXT NULL",
  "ALTER TABLE listing_vetoes ADD COLUMN IF NOT EXISTS brand TEXT NULL",
  "ALTER TABLE listing_vetoes ADD COLUMN IF NOT EXISTS size TEXT NULL",
  "ALTER TABLE listing_vetoes ADD COLUMN IF NOT EXISTS price_ron DOUBLE PRECISION NULL",
  "ALTER TABLE listing_vetoes ADD COLUMN IF NOT EXISTS value_band TEXT NULL",
  "ALTER TABLE listing_vetoes ADD COLUMN IF NOT EXISTS deal_score INT NULL",
  "ALTER TABLE listing_vetoes ADD COLUMN IF NOT EXISTS title TEXT NULL",
]

function databaseUrl() {
  return (
    process.env.DATABASE_URL ||
    process.env.COCKROACH_DATABASE_URL ||
    ""
  ).trim() || null
}

function sslConfig() {
  return { rejectUnauthorized: false }
}

function normalizeStatus(status) {
  if (status == null) return null
  const st = String(status)
  if (st === STATUS_HIDDEN_LEGACY) return STATUS_REMOVED
  return st
}

function coerceWriteStatus(status) {
  const st = normalizeStatus(status)
  if (!VALID.has(st)) {
    const err = new Error("invalid_status")
    err.status = 400
    throw err
  }
  return st
}

function coerceEnrichment(enrichment) {
  const out = {}
  for (const key of ENRICHMENT_FIELDS) out[key] = null
  if (!enrichment || typeof enrichment !== "object") return out
  for (const key of ENRICHMENT_FIELDS) {
    if (enrichment[key] == null) continue
    if (key === "price_ron") {
      const n = Number(enrichment[key])
      out[key] = Number.isFinite(n) ? n : null
    } else if (key === "deal_score") {
      const n = Number(enrichment[key])
      out[key] = Number.isFinite(n) ? Math.trunc(n) : null
    } else {
      const s = String(enrichment[key]).trim()
      out[key] = s || null
    }
  }
  if (!out.hunt_family && out.hunt_name) {
    out.hunt_family = resolveFamily(out.hunt_name)
  }
  return out
}

function itemId(rowOrId) {
  const raw =
    rowOrId && typeof rowOrId === "object"
      ? rowOrId.id ?? rowOrId.item_id
      : rowOrId
  if (raw == null) return null
  const n = Number(raw)
  return Number.isFinite(n) ? n : null
}

function statusFor(vetoes, rowOrId) {
  const id = itemId(rowOrId)
  if (id == null) return null
  return normalizeStatus(vetoes[id] || vetoes[String(id)] || null)
}

function isRemoved(vetoes, rowOrId) {
  return statusFor(vetoes, rowOrId) === STATUS_REMOVED
}

function isParked(vetoes, rowOrId) {
  return statusFor(vetoes, rowOrId) === STATUS_PARKED
}

function isBought(vetoes, rowOrId) {
  return statusFor(vetoes, rowOrId) === STATUS_BOUGHT
}

const isHidden = isRemoved

function deskRank(st) {
  if (st === STATUS_PARKED) return 1
  if (st === STATUS_BOUGHT) return 2
  return 0
}

function applyToFinds(rows, vetoes, { mode = "active" } = {}) {
  if (!VALID_MODES.has(mode)) {
    throw new Error(`unknown veto mode: ${mode}`)
  }
  const out = []
  for (const row of rows || []) {
    const st = statusFor(vetoes, row)
    if (st === STATUS_REMOVED) continue
    if (mode === "parked" && st !== STATUS_PARKED) continue
    if (mode === "bought" && st !== STATUS_BOUGHT) continue
    if (mode === "active" && st === STATUS_BOUGHT) continue
    const tagged = { ...row }
    if (st) tagged.veto_status = st
    else delete tagged.veto_status
    out.push(tagged)
  }
  return out.sort((a, b) => deskRank(a.veto_status) - deskRank(b.veto_status))
}

function applyToBundles(rows, vetoes, { mode = "active" } = {}) {
  if (!VALID_MODES.has(mode)) {
    throw new Error(`unknown veto mode: ${mode}`)
  }
  const out = []
  for (const bundle of rows || []) {
    const items = Array.isArray(bundle.items) ? bundle.items : []
    let kept = []
    for (const it of items) {
      const st = statusFor(vetoes, it)
      if (st === STATUS_REMOVED) continue
      if (st === STATUS_BOUGHT && (mode === "active" || mode === "parked")) continue
      const tagged = { ...it }
      if (st) tagged.veto_status = st
      else delete tagged.veto_status
      kept.push(tagged)
    }
    if (mode === "bought") {
      kept = kept.filter((it) => it.veto_status === STATUS_BOUGHT)
      if (kept.length < 1) continue
    } else if (kept.length < 2) {
      continue
    }
    if (
      mode === "parked" &&
      !kept.some((it) => it.veto_status === STATUS_PARKED)
    ) {
      continue
    }

    const row = { ...bundle, items: kept }
    let listingSum = 0
    for (const it of kept) {
      const p = Number(it.price)
      if (Number.isFinite(p)) listingSum += p
    }
    row.listing_sum = listingSum
    if (row.checkout_extra_ron != null) {
      const extra = Number(row.checkout_extra_ron)
      if (Number.isFinite(extra)) row.checkout_total = listingSum + extra
    }
    if (kept.some((it) => it.veto_status === STATUS_PARKED)) {
      row.veto_status = STATUS_PARKED
    } else if (kept.some((it) => it.veto_status === STATUS_BOUGHT)) {
      row.veto_status = STATUS_BOUGHT
    } else {
      delete row.veto_status
    }
    out.push(row)
  }
  return out.sort((a, b) => deskRank(a.veto_status) - deskRank(b.veto_status))
}

async function withClient(fn) {
  const url = databaseUrl()
  if (!url) return null
  const { Client } = pg
  let connectionString = url
  if (!fs.existsSync(path.join(os.homedir(), ".postgresql", "root.crt"))) {
    connectionString = url.replace(/sslmode=verify-full/gi, "sslmode=require")
  }
  const client = new Client({
    connectionString,
    ssl: sslConfig(),
    connectionTimeoutMillis: 8000,
    query_timeout: 15000,
  })
  try {
    await client.connect()
    await client.query(DDL)
    for (const stmt of ALTER_COLUMNS_SQL) {
      try {
        await client.query(stmt)
      } catch (err) {
        console.error("listingVetoes alter note:", err.message || err)
      }
    }
    try {
      await client.query(MIGRATE_HIDDEN_SQL)
    } catch (err) {
      console.error("listingVetoes migrate note:", err.message || err)
    }
    return await fn(client)
  } catch (err) {
    console.error("listingVetoes:", err.message || err)
    return null
  } finally {
    try {
      await client.end()
    } catch {
      /* ignore */
    }
  }
}

async function loadVetoMap() {
  const map = await withClient(async (client) => {
    const res = await client.query("SELECT item_id, status FROM listing_vetoes")
    const out = {}
    for (const row of res.rows) {
      out[Number(row.item_id)] = normalizeStatus(String(row.status))
    }
    return out
  })
  return map || {}
}

async function setVetoStatus(itemId, status, enrichment) {
  const id = Number(itemId)
  if (!Number.isFinite(id)) {
    const err = new Error("invalid_item_id")
    err.status = 400
    throw err
  }
  const st = coerceWriteStatus(status)
  const enr = coerceEnrichment(enrichment)
  const ok = await withClient(async (client) => {
    await client.query(
      `INSERT INTO listing_vetoes (
         item_id, status, updated_at,
         hunt_name, hunt_family, brand, size, price_ron, value_band, deal_score, title
       )
       VALUES ($1, $2, now(), $3, $4, $5, $6, $7, $8, $9, $10)
       ON CONFLICT (item_id) DO UPDATE SET
         status = EXCLUDED.status,
         updated_at = EXCLUDED.updated_at,
         hunt_name = COALESCE(EXCLUDED.hunt_name, listing_vetoes.hunt_name),
         hunt_family = COALESCE(EXCLUDED.hunt_family, listing_vetoes.hunt_family),
         brand = COALESCE(EXCLUDED.brand, listing_vetoes.brand),
         size = COALESCE(EXCLUDED.size, listing_vetoes.size),
         price_ron = COALESCE(EXCLUDED.price_ron, listing_vetoes.price_ron),
         value_band = COALESCE(EXCLUDED.value_band, listing_vetoes.value_band),
         deal_score = COALESCE(EXCLUDED.deal_score, listing_vetoes.deal_score),
         title = COALESCE(EXCLUDED.title, listing_vetoes.title)`,
      [
        id,
        st,
        enr.hunt_name,
        enr.hunt_family,
        enr.brand,
        enr.size,
        enr.price_ron,
        enr.value_band,
        enr.deal_score,
        enr.title,
      ],
    )
    return true
  })
  if (!ok) {
    const err = new Error("veto_db_unavailable")
    err.status = 503
    throw err
  }
  return { item_id: id, status: st }
}

async function clearVeto(itemId) {
  const id = Number(itemId)
  if (!Number.isFinite(id)) {
    const err = new Error("invalid_item_id")
    err.status = 400
    throw err
  }
  const ok = await withClient(async (client) => {
    await client.query("DELETE FROM listing_vetoes WHERE item_id = $1", [id])
    return true
  })
  if (!ok) {
    const err = new Error("veto_db_unavailable")
    err.status = 503
    throw err
  }
  return { item_id: id, cleared: true }
}

export {
  STATUS_REMOVED,
  STATUS_PARKED,
  STATUS_BOUGHT,
  STATUS_HIDDEN_LEGACY,
  databaseUrl,
  isRemoved,
  isHidden,
  isParked,
  isBought,
  applyToFinds,
  applyToBundles,
  loadVetoMap,
  setVetoStatus,
  clearVeto,
  normalizeStatus,
  coerceEnrichment,
}
